src/beatree/leaf/store.rs

```diff
@@ -0,0 +1,225 @@
+    io_sender: Sender<IoCommand>,
+    io_receiver: Receiver<CompleteIo>,
+    // monotonic page number, used when the free list is empty
+    nonce_page_number: PageNumber,
+    // this is the max supported page number, after which write fails,
+    // `ftruncate` of the file, to make it bigger, must be handled by upper layers
+    max_page_number: PageNumber,
+    free_list: FreeList,
+}
+
+impl LeafStore {
+    /// creates a LeafStore over a possibly already existing File
+    pub fn create(
+        fd: File,
+        free_list_head: Option<PageNumber>,
+        nonce_page_number: PageNumber,
+        io_handle_index: usize,
+        io_sender: Sender<IoCommand>,
+        io_receiver: Receiver<CompleteIo>,
+    ) -> LeafStore {
+        let file_size = fd
+            .metadata()
+            .expect("Error extracting metadata from LeafStore file")
+            .size() as usize;
+
+        let max_page_number = PageNumber((file_size / PAGE_SIZE) as u32);
+
+        Self {
+            free_list: FreeList::new(
+                &fd,
+                &io_sender,
+                io_handle_index,
+                &io_receiver,
+                free_list_head,
+            ),
+            store_file: fd,
+            io_handle_index,
+            io_sender,
+            io_receiver,
+            nonce_page_number,
+            max_page_number,
+        }
+    }
+
+    /// Returns the leaf page with the specified page number.
+    pub fn query(&self, pn: PageNumber) -> LeafPage {
+        let page = Box::new(Page::zeroed());
+
+        let mut command = Some(IoCommand {
+            kind: IoKind::Read(self.store_file.as_raw_fd(), pn.0 as u64, page),
+            handle: self.io_handle_index,
+            user_data: 0,
+        });
+
+        while let Some(c) = command.take() {
+            match self.io_sender.try_send(c) {
+                Ok(()) => break,
+                Err(TrySendError::Disconnected(_)) => panic!("I/O leaf store worker dropped"),
+                Err(TrySendError::Full(c)) => {
+                    command = Some(c);
+                }
+            }
+        }
+
+        // wait for completion
+        let completion = self
+            .io_receiver
+            .recv()
+            .expect("I/O leaf store worker dropped");
+        assert!(completion.result.is_ok());
+        let page = completion.command.kind.unwrap_buf();
+
+        LeafPage { inner: page }
+    }
+
+    // create a LeafStoreTx able to append and release leaves from the LeafStore
+    pub fn start_tx<'a>(&'a mut self) -> LeafStoreTx {
+        LeafStoreTx {
+            nonce_page_number: &mut self.nonce_page_number,
+            max_page_number: self.max_page_number,
+            free_list: &mut self.free_list,
+            to_allocate: vec![],
+            released: vec![],
+            exceeded: vec![],
+        }
+    }
+
+    // updates the mex supported page number based on the leaf store file size
+    pub fn update_size(&mut self) {
```

Comment:
When is this function supposed to be used? I had an impression that we should `stat` only upon initialization